### PR TITLE
Unblock Llama2 ONNX export w/ sdpa by falling back to manual impl

### LINKED ISSUE #28610: ONNX export failure for models invoking SDPA attention
> ValueError: Attention using SDPA can not be traced with torch.jit.trace when no attention_mask is provided. To solve this issue, please either load your model with the argument `attn_implementation="eager"` or pass an attention_mask input when tracing the model. 

There has been some discussion about its possible resolutions in the ONNX exporter team. I'd like to post an issue here as well to seek advice and preferences.

1. Check `torch.jit.is_tracing()` and fallback to eager attn implementation if needed.
2. Create `attention_mask` before passing to SDPA if it is None.
3. Support SDPA tracing w/o attention_mask (not sure how feasible this is).

### PR DESCRIPTION
**Pull request by BowenBao**
# What does this PR do?

Unblocks Llama2 ONNX export with sdpa by falling back to manual implementation.

> ValueError: Attention using SDPA can not be traced with torch.jit.trace when no attention_mask is provided. To solve this issue, please either load your model with the argument attn_implementation="eager" or pass an attention_mask input when tracing the model.

Fixes #28610

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@fxmarty

### CODE REVIEW

**Review by thiagocrepaldi — APPROVED** [review]
LGTM. Maybe add a unit test for the torch.jit.trace case?

**Review by ArthurZucker — COMMENTED** [review]
Thanks for this! It's gonna be a bit hard to merge this. Would you mind checking if #27931 fixes the issue? It shall be merged before and should simplify all of that logic

**Comment by BowenBao** [comment]
Hi @ArthurZucker, I have validated the issue is fixed under your PR, thanks! Do you have an ETA when it will get merged? Our workstreams have been blocked by this issue for a while, we need to resolve this export issue asap.

**Comment by ArthurZucker** [comment]
This week 😉 Waiting for @gante's green light and will merge it

File: src/transformers/models/llama/modeling_llama.py
Comment (fxmarty) [inline src/transformers/models/llama/modeling_llama.py:676]: This means that we call torch.jit.is_tracing as many times as there are layers.

**Comment by fxmarty** [comment]
I don't understand why this change is necessary. The error that is normally raised
```
ValueError: Attention using SDPA can not be traced with torch.jit.trace when no attention_mask is provided. To solve this issue, please either load your model with the argument attn_implementation="eager" or pass an attention_mask input when tracing the model.
```
explicitly gives a solution.

**Comment by thiagocrepaldi** [comment]
@ArthurZucker @BowenBao I believe we can close this issue now that #27931 was merged

**Pull request closed without merging**
